### ISSUE TITLE
propose-upstream does not adjust the spec's Version:

### ISSUE THREAD

**Issue by martinpitt** [packit/packit-service#1515]
We tried to use packit for Fedora with cockpit, using [this packit.yaml](https://github.com/cockpit-project/cockpit/blob/3d8af4cd125293c5c258c48209606a258a22d1a7/packit.yaml). Aside from the already filed hacks (#1505 and https://github.com/packit/packit/issues/1560), we have this hack:
```
   post-upstream-clone:
    # packit will overwrite the version in its "fix spec file" stage
    - tools/create-spec --version 0 --build-all -o cockpit.spec tools/cockpit.spec.in
```

We need this because our git repo does not contain a cockpit.spec, it needs to be built from a template. This works fine for COPR builds, but not for propose_downstream -- the release [failed](https://dashboard.packit.dev/results/propose-downstream/597) like this:
```
WARNING Version '0' in spec file is outdated
[...]
DEBUG  Upstream archive name: cockpit-0.tar.xz
DEBUG  Archive 'cockpit-0.tar.xz' not found in the lookaside cache.
```

So apparently that "fix spec file" magic does not get applied to propose_downstream stage?

*Ideally* packit would take cockpit.spec from the downloaded release tarball -- everything is correct there. I tried to drop post-upstream-clone, and propose_downstream immediately failed with "no spec file", so that doesn't work.

After some 10 other approaches which failed, I now have this hack:
```diff
--- packit.yaml
+++ packit.yaml
@@ -1,8 +1,8 @@
 upstream_project_url: https://github.com/martinpitt/cockpit
 actions:
   post-upstream-clone:
-    # packit will overwrite the version in its "fix spec file" stage
-    - tools/create-spec --version 0 --build-all -o cockpit.spec tools/cockpit.spec.in
+    # HACK: packit's "fix spec file" stage does not kick in here, and it does not take the spec from the tarball
+    - sh -exc 'tools/create-spec --version $(git describe) --build-all -o cockpit.spec tools/cockpit.spec.in'
   create-archive:
     # The sandcastle doesn't have enough ram to run webpack, so wait
     # until the webpack-jumpstart workflow has run and grab the result.
```

to build the checkout's cockpit.spec with the expected version number.

Even with that, the [propose-downstream log](https://dashboard.packit.dev/results/propose-downstream/650) still has no sign of actually downloading the cockpit-269.2.tar.xz tarball from https://github.com/martinpitt/cockpit/releases , but at least it has the correct version number. It auto-retried a few times ("tarball not yet available") and then failed. But that might be an artifact of cockpit.spec having a Source0: pointing to cockpit-project/cockpit insteaad of my fork. (I'll try to hack this next).

So in summary, I found it really difficult to get propose-upstream working in this scenario (spec file needs to be built). Is there some best practice here?

**Comment by TomasTomecek** [comment]
I filed https://github.com/packit/packit-service/issues/223 in the past, although your issue is very concrete.

The best practice so far is to manually adjust `Version` when doing releases. Which won't work for you since you generate the spec file.

If this `--version $(git describe)` would work for everyone, that would be a win.

**Comment by martinpitt** [comment]
> Would running a fix-spec-file action during propose-downstream work for you?

I think so -- at least the builtin default does the right thing for COPR, i.e. replace the spec's `Version:` with the actual version. Packit already knows this, but does not expose it as a variable (`$PACKIT_PROJECT_VERSION` is not set for post-upstream-clone). As I said, this is not my most liked solution, but it would help to avoid these nasty shell/seddery hacks. FTR, I updated the description with a fixed `git` command, the previous one didn't work.

> I am not sure if this can be used for all users -- upstream archive does not need to contain a spec-file at all.

Right, but *if* it contains it, then I feel this should be preferred to the version in the cloned git? These days, most upstream projects don't keep the git-tracked spec up to date with the version, that's outright impossible with COPRs from PRs even. Do you think it would be possible to apply the normal "search for a .spec file" approach in the upstream tarball for propose-downstream first, and only then fall back to all the `post-upstream-clone:`, `fix-spec-file` etc. magic?

**Comment by lachmanfrantisek** [comment]
OK, so we have basically three action items:

* [ ] update the spec-version during a `propose-downstream` job.
  * What about doing this as a `fix-spec-file` action? Or a new `bump-spec-version` so users can disable this?
  * Do we want to use a new version (get from tag) or take the highest from tag and spec-file? Also, we can introduce `get-current-version`.
  * This might help our tools as well so we don't need to maintain the version in the upstream specfile...
* [ ] Provide `PACKIT_PROJECT_VERSION` during propose-downstream.
  * We can do that if it makes sense. (With service, it will be a version transformed from git tag using the `upstream_tag_template`.)
* [ ] Try getting a spec-file from the source archive.
  * But we are getting the info about the archive from the spec-file...;(
  * We can possibly do something with that but I am worried this can mess the already-a-bit-complicated process more.

**Comment by lachmanfrantisek** [comment]
Last update of the action items:
(Leaving the spec-from-archive for later):

* [ ] update the spec-version during a propose-downstream job.
  * Do this as a `bump-spec-version` action so users can disable this.
  * Use the value from `get-current-version` action or get it from the tag by default.
  * Preserve the warning if upstream-release-monitoring has a new version.
 * [ ] provide `PACKIT_PROJECT_VERSION` env.var. during `pre-sync` action. (As a value we use the output from the `get-current-version` action or tag by default. (What we use in spec-file.)
 * [ ] make sure the following use-cases work:
   * [ ] propose-downstream in service with the spec-file present in upstream (e.g. Packit; spec-file does not need to updated in upstream)
   * [ ] propose-downstream in service with the spec-file generated during action (e.g. Anaconda, Cockpit)
   * [ ] propose-downstream run via CLI for both upstream-located and action-generated specfile

**Comment by csomh** [comment]
>   * Do this as a `bump-spec-version` action so users can disable this.

Maybe call this `set-spec-version`, as versions are not only bumped.

Figure out some way, so that by default this actions doesn't do anything. This way we could maintain backwards compatibility.

**Comment by lachmanfrantisek** [comment]
Regarding the syntax, we have multiple options:

* Nothing by default, the default behaviour when `true` yaml value is configured, user-defined command otherwise.
  *  Here, I am worried about a collision between a dummy command (like `true`) and some special value (`true`/`1`/`True`/`yes`/..). Maybe, we can have some explicit special value like `packit-default`.
* Have a dedicated config flag so turn this on. (Nothing set means no change, flag set means the action is run with the default to use the version from `get-current-version`/tag.)

The person who will work on this will see what is better to use.

**Comment by lbarcziova** [comment]
Hi, I was looking into the `propose_downstream` code base and found out that we actually do set the `Version` in the specfile ([here](https://github.com/packit/packit/blob/63407fe7a6c7a97a678893caa5c6eae7da2991e9/packit/utils/changelog_helper.py#L92)), in case of Packit Service we set here the version that we get from the upstream tag corresponding to the release that triggered the `propose_downstream`, but we do it after running the `post_upstream_clone` action, when preparing files to sync. 

If I understand correctly, the real problem here is that the archive we try to get is not correct. But checking how you create the specfile in `post_upstream_clone`, the `Source0` is hardcoded after running the action and since we set the `Version` in specfile correctly later, it is not considered when downloading the sources because the source name is already hardcoded. 

Would it be possible for you to use in Source0 the version macro maybe? Or do you still think providing something similar to  `fix-spec-file` for `propose_downstream` where you can update the `Source0` would be the best option?

**Comment by martinpitt** [comment]
> Would it be possible for you to use in Source0 the version macro maybe?

Absolutely, yes, thanks for spotting! That also explains why we don't run into this problem in e.g. [c-podman](https://github.com/cockpit-project/cockpit-podman/blob/main/packit.yaml). Its [spec file](https://github.com/cockpit-project/cockpit-podman/blob/main/packaging/cockpit-podman.spec.in#L25) uses the version macro in the way you mention:

    Source0:        https://github.com/cockpit-project/%{name}/releases/download/%{version}/%{name}-%{version}.tar.xz

and does the upstream-side templating through this indirection:

    Version:        %{VERSION}

I'll try the same for cockpit. 

If that works, I think we can close this -- indeed my initial "Ideally packit would take cockpit.spec from the downloaded release tarball -- everything is correct there" proposal is difficult due to the chicken-and-egg problem.

Thanks @lbarcziova !

**Comment by lbarcziova** [comment]
Great! Let me know if there will be any problems after this change and in that case we can still implement some action.

**Comment by martinpitt** [comment]
@lbarcziova ! Works! https://github.com/cockpit-project/cockpit/pull/17450 Thanks for your help!
